Add MARS (Import only)

kirill-sizov opened the pull request:
<!-- Contributing guide: https://github.com/openvinotoolkit/datumaro/blob/develop/CONTRIBUTING.md -->

### Summary
<!--
Resolves #111 and #222.
Depends on #1000 (for series of dependent commits).

This PR introduces this capability to make the project better in this and that.

- Added this feature
- Removed that feature
- Fixed the problem #1234
-->

### How to test
<!-- Describe the testing procedure for reviewers, if changes are
not fully covered by unit tests or manual testing can be complicated. -->

### Checklist
<!-- Put an 'x' in all the boxes that apply -->
- [ ] I submit my changes into the `develop` branch
- [ ] I have added description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/datumaro/blob/develop/CHANGELOG.md)
- [x] I have updated the [documentation](
  https://github.com/openvinotoolkit/datumaro/tree/develop/docs) accordingly
- [x] I have added tests to cover my changes
- [ ] I have [linked related issues](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

### License

- [ ] I submit _my code changes_ under the same [MIT License](
  https://github.com/openvinotoolkit/datumaro/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2021 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

zhiltsov-max commented on tests/test_mars_format.py:
```diff
@@ -0,0 +1,60 @@
+        expected_dataset = Dataset.from_iterable([
+            DatasetItem(id='0001C1T0001F001', image=np.ones((10, 10, 3)),
+                subset='train', annotations=[Label(label=2, attributes={
+                    'is_distractors': False,
+                    'is_junk': False,
+                    'pedestrian_id': '0001',
```
These 3 attributes look mutually exclusive. Can we reduce them to only one?

IRDonch replied:
Does the value of this attribute represent the same information as the label value? If so, why does it exist?

zhiltsov-max commented on tests/test_mars_format.py:
```diff
@@ -0,0 +1,60 @@
+                    'pedestrian_id': '0001',
+                    'camera_id': '1',
+                    'track_id': '0001',
+                    'frame_id': '001'})]
```
All the label attributes here look like item attributes. Why do you think we should store them in annotations?

zhiltsov-max commented on tests/test_mars_format.py:
```diff
@@ -0,0 +1,60 @@
+                    'track_id': '0081',
+                    'frame_id': '201'})]
+            ),
+        ], categories=['00-1', '0000', '0001'])
```
The LFW, WiderFace and MARS are related to the same task, consider aligning their representations.

kirill-sizov replied:
What do you mean by "aligning"? Do you mean that the formats should have a similar representation in Datumaro or something else?

zhiltsov-max replied:
Yes, exactly. It will look pretty hard to use if these formats are not convertible between each other.

zhiltsov-max commented on datumaro/plugins/mars_format.py:
```diff
@@ -0,0 +1,121 @@
+
+                if not fnmatch.fnmatch(image_name,
+                        label + MarsPath.IMAGE_NAME_POSTFIX):
+                    log.warning(f'The image {image_path} will be skip because '
```
```suggestion
                    log.warning(f'The image {image_path} will be skipped because '
```

kirill-sizov commented on tests/test_mars_format.py:
```diff
@@ -0,0 +1,60 @@
+    def test_can_import(self):
+        expected_dataset = Dataset.from_iterable([
+            DatasetItem(id='0001C1T0001F001', image=np.ones((10, 10, 3)),
+                subset='train', annotations=[Label(label=2, attributes={
```
Because each image in the dataset represents itself as clipped bounding box, and we don't have original image or information about coordinates of this bounding box in the original image.

zhiltsov-max replied:
Probably, I don't quite understand the problem. We have an image, which is a bbox cut from the original image. But original images are not used in this format, so this bbox should just fit the whole image, shouldn't it?

kirill-sizov replied:
If I understand you correctly,  in such implementation we need to know size of image to assign the correct values to the coordinates of the bbox. Would there be an overhead of reading the image to find out the size of the image?

zhiltsov-max replied:
It's a fair point. Yes, it can affect loading time, however, if we look at masks in VOC-like formats, it works pretty fine even with more complex processing. I think, we can try to implement this.

zhiltsov-max commented on site/content/en/docs/formats/mars.md:
````diff
@@ -0,0 +1,76 @@
+
+```
+mars_dataset
+├── <bbox_subset_name1>
````
Will this format support metafile?

kirill-sizov replied:
I'm not sure that this format needs a metafile.

zhiltsov-max replied:
Probably, I can agree.

IRDonch commented on site/content/en/docs/formats/mars.md:
```diff
@@ -0,0 +1,76 @@
+MARS is a dataset for and motion analysis and person identification task,
+and this dataset it's extension of Market-1501 dataset format.
+MARS dataset is available for downloading
+[here](http://zheng-lab.cecs.anu.edu.au/Project/project_mars.html)
```
```suggestion
[here](https://zheng-lab.cecs.anu.edu.au/Project/project_mars.html)
```

IRDonch commented on site/content/en/docs/formats/mars.md:
```diff
@@ -0,0 +1,76 @@
+
+## Format specification
+
+MARS is a dataset for and motion analysis and person identification task,
```
```suggestion
MARS is a dataset for the motion analysis and person identification task,
```

IRDonch commented on site/content/en/docs/formats/mars.md:
```diff
@@ -0,0 +1,76 @@
+## Format specification
+
+MARS is a dataset for and motion analysis and person identification task,
+and this dataset it's extension of Market-1501 dataset format.
```
There seems to be no verb here.

IRDonch commented on site/content/en/docs/formats/mars.md:
```diff
@@ -0,0 +1,76 @@
+- `Label`
+
+Required attributes:
+- `is_distractors` (bool): True when image with distractors,
```
"When" implies a dependent clause, which needs a verb - but there is no verb here.

IRDonch commented on site/content/en/docs/formats/mars.md:
```diff
@@ -0,0 +1,76 @@
+Required attributes:
+- `is_distractors` (bool): True when image with distractors,
+    which negatively affect retrieval accuracy
+- `is_junk`: True for junk image which do not affect retrieval accuracy;
```
```suggestion
- `is_junk`: True for junk images which do not affect retrieval accuracy;
```

Also, what's the data type?

IRDonch commented on site/content/en/docs/formats/mars.md:
```diff
@@ -0,0 +1,76 @@
+- `is_distractors` (bool): True when image with distractors,
+    which negatively affect retrieval accuracy
+- `is_junk`: True for junk image which do not affect retrieval accuracy;
+- `pedestrian_id`: four-digit number in format `%04d`;
```
I'm confused by this. First of all, isn't it redundant to say `%04d` when you already said it's a four-digit number? Second, what data type is this attribute supposed to have? If `int`, then the format doesn't make sense. If `str`, then why not `int`?

In addition, the meaning of the attribute should be explained beyond just "number".

IRDonch commented on site/content/en/docs/formats/mars.md:
````diff
@@ -0,0 +1,76 @@
+├── ...
+```
+
+All images in MARS dataset has strict convention of naming:
````
```suggestion
All images in MARS dataset follow a strict convention of naming:
```

IRDonch commented on site/content/en/docs/formats/mars.md:
````diff
@@ -0,0 +1,76 @@
+xxxxCxTxxxxFxxx.jpg
+```
+- the first four digits indicate the pedestrian's number;
+- digits after `C` indicate the camera id;
````
```suggestion
- digit after `C` indicates the camera id;
```

(Since there's only one digit.)

IRDonch commented on site/content/en/docs/formats/mars.md:
```diff
@@ -0,0 +1,76 @@
+- the first four digits indicate the pedestrian's number;
+- digits after `C` indicate the camera id;
+- four digits after `T` indicate the track id for this pedestrian;
+- three digits after `F` indicate the frame id with this track
```
```suggestion
- three digits after `F` indicate the frame id with this track.
```

IRDonch commented on site/content/en/docs/formats/mars.md:
```diff
@@ -0,0 +1,76 @@
+- three digits after `F` indicate the frame id with this track
+
+> Note: there are two specific pedestrian IDs 0000 and 00-1
+> which indicate distracting images and unwanted images respectively
```
```suggestion
> which indicate distracting images and unwanted images respectively.
```

kirill-sizov commented:
I fixed all comments and updated `market1501` format for compatibility with `mars` format, there is changes that I made for `market1501`:
- **Importer**: made heuristic more unique
- **Extractor**: made `query` items as separate subset, because the original dataset don't suppose that query image relate with images from other subsets
- **Load items**: added attributes according original dataset [description](http://zheng-lab.cecs.anu.edu.au/Project/project_reid.html) (I used `track_id` instead of `sequence_id` because these two attributes have the same meaning)
- **Converter**: implemented this strategy: if item has `person_id` attribute `Converter` try to save this item with original image naming convention, using default values for unknown attributes and incrementing `frame_id` for collisions.

I decided to keep both `person_id` attribute and `Label` for MARS format for compatibility with classification-task formats.
I also used the `str` type for `person_id`, for cases where users would like to mark some items with a special identifier.

zhiltsov-max commented on tests/test_market1501_format.py:
```diff
@@ -120,20 +101,16 @@ def test_can_save_dataset_with_no_save_images(self):
-                    'query': True
-                }),
+            DatasetItem(id='0001_c1s1_0000_00', image=Image(
+                    path='0001_c1s1_0000_00.JPEG', data=np.zeros((4, 3, 3))),
```
I see the new test doesn't check for relative path support. Are they supported?

zhiltsov-max commented on datumaro/plugins/market1501_format.py:
```diff
@@ -3,137 +3,148 @@
-                        dirname = Market1501Path.QUERY_DIR
+                pid = item.attributes.get('person_id')
+                match = Market1501Path.PATTERN.fullmatch(item.id)
+                if not match and pid is not None:
```
```suggestion
                if not match and pid:
```
I'm not sure we can use empty `pid`

zhiltsov-max reviewed (changes requested):
Looks good, please update Market docs.
